tsl/src/bgw_policy/continuous_aggregate_api.c

```diff
@@ -39,3 +54,236 @@ policy_continuous_aggregate_proc(PG_FUNCTION_ARGS)
 
 	PG_RETURN_VOID();
 }
+
+static Oid
+ts_cagg_permissions_check(Oid cagg_oid, Oid userid)
+{
+	Oid ownerid = ts_rel_get_owner(cagg_oid);
+
+	if (!has_privs_of_role(userid, ownerid))
+		ereport(ERROR,
+				(errcode(ERRCODE_INSUFFICIENT_PRIVILEGE),
+				 errmsg("must be owner of continuous aggregate \"%s\"", get_rel_name(cagg_oid))));
+
+	return ownerid;
+}
+
+static void
+json_add_dim_interval_value(JsonbParseState *parse_state, const char *json_label, Oid dim_type,
+							Datum value)
+{
+	switch (dim_type)
+	{
+		case INTERVALOID:
+			ts_jsonb_add_interval(parse_state, json_label, DatumGetIntervalP(value));
+			break;
+		case INT2OID:
+			ts_jsonb_add_int64(parse_state, json_label, DatumGetInt16(value));
+			break;
+		case INT4OID:
+			ts_jsonb_add_int64(parse_state, json_label, DatumGetInt32(value));
+			break;
+		case INT8OID:
+			ts_jsonb_add_int64(parse_state, json_label, DatumGetInt64(value));
+			break;
+		default:
+			ereport(ERROR,
+					(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+					 errmsg("unsupported datatype for value:%s, expected type: %s",
+							format_type_be(value),
+							format_type_be(dim_type))));
+	}
+}
+
+Datum
+policy_refresh_cagg_add(PG_FUNCTION_ARGS)
+{
+	NameData application_name;
+	NameData refresh_name;
+	NameData proc_name, proc_schema, owner;
+	Cache *hcache;
+	Hypertable *mat_ht;
+	Dimension *dim;
+	ContinuousAgg *cagg;
+	int32 job_id, mat_htid;
+	Datum start_interval, end_interval;
+	Interval refresh_interval;
+	Oid dim_type, start_interval_type, end_interval_type;
+	Oid cagg_oid, owner_id;
+	List *jobs;
+	bool if_not_exists;
+	if (PG_ARGISNULL(1) || PG_ARGISNULL(2) || PG_ARGISNULL(3))
+	{
+		ereport(ERROR,
+				(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+				 errmsg("cannot add refresh policy due to null argument values"),
```

Comment:
If we are to check these parameter values I would check them each individually, e.g.:

```suggestion
				(errcode(ERRCODE_INVALID_PARAMETER_VALE),
				 errmsg("cannot specify NULL start interval").
				 ...
```